[dynamo] Deprecate enable_cpp_framelocals_guard_eval config variable - default: True

BartlomiejStemborowski opened the pull request:
[dynamo] Deprecate enable_cpp_framelocals_guard_eval config variable - default: True

Reading the feature enabling param `enable_cpp_framelocals_guard_eval `at the CPP level is time consuming and slows down the operation of the dynamo as it is done every time the function using this param is called. Reading the value only once at init isn’t an option as it would disable the modification of this param at the runtime. Since this feature is enabled by default for some time and it doesn’t cause known issues, the `enable_cpp_framelocals_guard_eval `configuration param will be deprecated by this commit and its value is hardcoded to true.

Local microbenchmark dynamo_guard_eval.py:
- 931.9 us -> 538.9 us (3.10)

@williamwen42 @jansel @anijain2305 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

BartlomiejStemborowski committed:
[dynamo] Read enable_cpp_framelocals_guard_eval config variable only once

Flow in the function run_root_guard_manager depends on the value of the enable_cpp_framelocals_guard_eval config
variable from the torch._dynamo.config file. Reading this variable each time is time consuming as it requires
importing whole module. In order to do it only once, the local variable enable_cpp_framelocals_guard_eval is
converted into a static one and is initialized with the value read from the config file.

Local microbenchmark dynamo_guard_eval.py:
- 931.9 us -> 553.7 us (3.10)

BartlomiejStemborowski commented:
@williamwen42 @jansel @anijain2305
Please help with the review

BartlomiejStemborowski commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -5224,9 +5224,10 @@ bool run_root_guard_manager(void* root, FrameLocalsMapping* f_locals) {
-  py::object config_module = py::module_::import("torch._dynamo.config");
-  bool enable_cpp_framelocals_guard_eval =
-      config_module.attr("enable_cpp_framelocals_guard_eval").cast<bool>();
+  static bool enable_cpp_framelocals_guard_eval =
+      py::module_::import("torch._dynamo.config")
+          .attr("enable_cpp_framelocals_guard_eval")
+          .cast<bool>();
```
Sure, I will try to implement the fix using CPython API and update the patch.
However, I am wondering if this `enable_cpp_framelocals_guard_eval` switch is still needed. As its default value is true and it seems to be just a feature switch maybe it is okay just to remove it and leave only this default code path.
CC
@williamwen42

williamwen42 reviewed (commented):
We had a bunch of issues in the past involving loading modules statically, which is why I'm loading every time (cpython also caches these imports, so we're not actually doing the full import every time).

I also haven't received any complaints about this feature recently, so we could just deprecate this config.

BartlomiejStemborowski committed:
[dynamo] Deprecate enable_cpp_framelocals_guard_eval config variable - default: True

Reading the feature enabling param enable_cpp_framelocals_guard_eval at the CPP level is time consuming
and slows down the operation of the dynamo as it is done every time the function using this param is called.
Reading the value only once at init isn’t an option as it would disable the modification of this param at
the runtime. Since this feature is enabled by default for some time and it doesn’t cause known issues,
the enable_cpp_framelocals_guard_eval configuration param will be deprecated by this commit and its value
is hardcoded to true.

BartlomiejStemborowski commented:
Updated the PR with the patch deprecating the `enable_cpp_framelocals_guard_eval ` config. Please check.
@williamwen42 @jansel @anijain2305

williamwen42 commented:
@pytorchbot merge
